ml-algorithms/src/main/java/org/opensearch/ml/engine/algorithms/agent/MLPlanExecuteAndReflectAgentRunner.java

```diff
@@ -0,0 +1,521 @@
+
+        params.put(NO_ESCAPE_PARAMS_FIELD, DEFAULT_NO_ESCAPE_PARAMS);
+
+        // setting defaults for llm response
+        if (params.containsKey(LLM_INTERFACE) && (!params.containsKey(LLM_RESPONSE_FILTER) || params.get(LLM_RESPONSE_FILTER).isEmpty())) {
+            String llmInterface = params.get(LLM_INTERFACE);
+            String llmResponseFilter = switch (llmInterface.trim().toLowerCase(Locale.ROOT)) {
+                case LLM_INTERFACE_BEDROCK_CONVERSE_CLAUDE, LLM_INTERFACE_BEDROCK_CONVERSE_DEEPSEEK_R1 ->
+                    "$.output.message.content[0].text";
+                case LLM_INTERFACE_OPENAI_V1_CHAT_COMPLETIONS -> "$.choices[0].message.content";
+                default -> throw new MLException(String.format("Unsupported llm interface: %s", llmInterface));
+            };
+
+            params.put(LLM_RESPONSE_FILTER, llmResponseFilter);
+        }
+    }
+
+    private void usePlannerPromptTemplate(Map<String, String> params) {
+        params.put(PROMPT_TEMPLATE_FIELD, PLANNER_PROMPT_TEMPLATE);
+        populatePrompt(params);
+    }
+
+    private void useRevalPromptTemplate(Map<String, String> params) {
+        params.put(PROMPT_TEMPLATE_FIELD, REFLECT_PROMPT_TEMPLATE);
+        populatePrompt(params);
+    }
+
+    private void usePlannerWithHistoryPromptTemplate(Map<String, String> params) {
+        params.put(PROMPT_TEMPLATE_FIELD, PLANNER_WITH_HISTORY_PROMPT_TEMPLATE);
+        populatePrompt(params);
+    }
+
+    private void populatePrompt(Map<String, String> allParams) {
+        String promptTemplate = allParams.get(PROMPT_TEMPLATE_FIELD);
+        StringSubstitutor promptSubstitutor = new StringSubstitutor(allParams, "${parameters.", "}");
+        String prompt = promptSubstitutor.replace(promptTemplate);
+        allParams.put(PROMPT_FIELD, prompt);
+    }
+
+    @Override
+    public void run(MLAgent mlAgent, Map<String, String> apiParams, ActionListener<Object> listener) {
+        Map<String, String> allParams = new HashMap<>();
+        allParams.putAll(apiParams);
+        allParams.putAll(mlAgent.getParameters());
+
+        setupPromptParameters(allParams);
+
+        // planner prompt for the first call
+        usePlannerPromptTemplate(allParams);
+
+        String memoryId = allParams.get(MEMORY_ID_FIELD);
+        String memoryType = mlAgent.getMemory().getType();
+        String appType = mlAgent.getAppType();
+        int messageHistoryLimit = DEFAULT_MESSAGE_HISTORY_LIMIT;
+
+        ConversationIndexMemory.Factory conversationIndexMemoryFactory = (ConversationIndexMemory.Factory) memoryFactoryMap.get(memoryType);
+        conversationIndexMemoryFactory
+            .create(apiParams.get(USER_PROMPT_FIELD), memoryId, appType, ActionListener.<ConversationIndexMemory>wrap(memory -> {
+                memory.getMessages(ActionListener.<List<Interaction>>wrap(interactions -> {
+                    List<String> completedSteps = new ArrayList<>();
```

Comment:
This is to add conversation history to completed steps in current plan. Suggest add conversation history to messages list with different roles. This will need some time to refactor and test. Add a todo for fixing this part?